server/util.go

```diff
@@ -280,13 +283,68 @@ func dumpBinaryRow(buffer []byte, columns []*ColumnInfo, row chunk.Row) ([]byte,
 	return buffer, nil
 }
 
-func dumpTextRow(buffer []byte, columns []*ColumnInfo, row chunk.Row) ([]byte, error) {
+type textDumper struct {
+	chsName  string
+	encoding charset.Encoding
+	buffer   []byte
+	isBinary bool
+	isNull   bool
+}
+
+// newTextDumper creates a new textDumper.
+func newTextDumper(chs string) *textDumper {
+	return &textDumper{
+		chsName:  chs,
+		encoding: *charset.NewEncoding(chs),
+		buffer:   nil,
+		isBinary: chs == charset.CharsetBinary,
+		isNull:   len(chs) == 0,
+	}
+}
+
+// clean prevent the textDumper from holding too much memory.
+func (d *textDumper) clean() {
+	d.buffer = nil
+}
+
+func (d *textDumper) charsetNeedDynUpdate() bool {
+	return d.isBinary || d.isNull
+}
+
+func (d *textDumper) encode(src []byte) []byte {
+	if !d.encoding.Enabled() {
+		return src
+	}
+	result, err := d.encoding.Encode(d.buffer, src)
+	if err != nil {
+		logutil.BgLogger().Debug("encode error", zap.Error(err))
+	}
+	return result
+}
+
+func dumpTextRow(buffer []byte, columns []*ColumnInfo, row chunk.Row, d *textDumper) ([]byte, error) {
+	if d == nil {
+		d = &textDumper{}
+	}
 	tmp := make([]byte, 0, 20)
+	if d.charsetNeedDynUpdate() {
```

Comment:
Could we test this logic？ Maybe we can add some columns in `TestDumpTextValue`.
Rest LGTM.